pycaret/internal/PycaretExperiment.py

```diff
@@ -16667,27 +16634,437 @@ def tune_model(
+
+        warnings.filterwarnings("ignore")
+
+        import logging
+
+        np.random.seed(self.seed)
+
+        self.logger.info("Copying training dataset")
+        # Storing X_train and y_train in data_X and data_y parameter
+        data_X = self.X_train.copy()
+        data_y = self.y_train.copy()
+
+        # Replace Empty DataFrame with None as empty DataFrame causes issues
+        if (data_X.shape[0] == 0) or (data_X.shape[1] == 0):
+            data_X = None
+
+        display.move_progress()
+
+        # setting optimize parameter
+
+        # TODO: Changed compared to other PyCaret UseCases (Check with Antoni)
+        # optimize = optimize.scorer
+        compare_dimension = optimize_container.display_name
+        optimize = {optimize: optimize_container}
+
+        # convert trained estimator into string name for grids
+
+        self.logger.info("Checking base model")
+
+        is_stacked_model = False
+
+        if hasattr(estimator, "final_estimator"):
+            self.logger.info("Model is stacked, using the definition of the meta-model")
+            is_stacked_model = True
+            estimator_id = self._get_model_id(estimator.final_estimator)
+        else:
+            estimator_id = self._get_model_id(estimator)
+        if estimator_id is None:
+            if custom_grid is None:
+                raise ValueError(
+                    "When passing a model not in PyCaret's model library, the custom_grid parameter must be provided."
+                )
+            estimator_name = self._get_model_name(estimator)
+            estimator_definition = None
+            self.logger.info("A custom model has been passed")
+        else:
+            estimator_definition = self._all_models_internal[estimator_id]  # Container
+            estimator_name = estimator_definition.name
+        self.logger.info(f"Base model : {estimator_name}")
+
+        # If no special tunable class is defined inside PyCaret then just clone the estimator
+        if estimator_definition is None or estimator_definition.tunable is None:
+            model = clone(estimator)
+        # If special tunable class is defined, then use that instead
+        else:
+            self.logger.info("Model has a special tunable class, using that")
+            model = clone(estimator_definition.tunable(**estimator.get_params()))
+        is_stacked_model = False
+
+        base_estimator = model
+
+        display.update_monitor(2, estimator_name)
+        display.display_monitor()
+
+        display.move_progress()
+
+        self.logger.info("Declaring metric variables")
+
+        """
+        MONITOR UPDATE STARTS
+        """
+
+        display.update_monitor(1, "Searching Hyperparameters")
+        display.display_monitor()
+
+        """
+        MONITOR UPDATE ENDS
+        """
+
+        self.logger.info("Defining Hyperparameters")
+
+        # TODO: Replace with time series specific code
+        def total_combintaions_in_grid(grid):
```

Comment:
We can use ParameterGrid from sklearn for getting the combinations, I believe it'd be the easiest way: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.ParameterGrid.html - and it will allow for easy random search too

Reply:
Hi @Yard1 , That is what I am using https://github.com/pycaret/pycaret/blob/568dd5365ade055398cbf3e5e0084d7fd4085d54/pycaret/time_series.py#L2435. I am not sure it has methods to compute number of combinations.

Reply:
I think it supports len. If not you should be able to calculate the number of possible combinations using some math, we probably don't need loops here

Reply:
Do you mean something like this? https://github.com/alan-turing-institute/sktime/blob/ee7a06843a44f4aaec7582d847e36073a9ab0566/sktime/forecasting/model_selection/_tune.py#L428

Reply:
Yes